### PR TITLE
Add amplitude parsing from zmachine

### PR DESCRIPTION
**Pull request by Shiro-Raven**

### CODE REVIEW

**Review by hsim13372 — COMMENTED** [review]
Looks good to me although we may want to also bring some relevant tests from zmachine?

**Comment by Shiro-Raven** [comment]
> Looks good to me although we may want to also bring some relevant tests from zmachine?

Turns out there are no tests either in z-machine nor in openfermion-psi4 where this function is originally from

File: src/python/qepsi4/_ccsd_amplitude_parser.py [hunk 1]
```diff
@@ -0,0 +1,303 @@
...
+    single_amps = []
+    double_amps = []
+
+    # Read T1's
+
+    if get_mp2_amplitudes:
+
+        # generate all possible singlet excitations
+
+        # for i in range(int(n_frozen_amplitudes), n_alpha_electrons):
```
Comment (mstechly) [inline src/python/qepsi4/_ccsd_amplitude_parser.py:166]: This looks like unnecessarily commented?

File: src/python/qepsi4/_ccsd_amplitude_parser.py [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+############################################################################
+
+import numpy as np
+from openfermion import FermionOperator
+
+# Original function: https://github.com/quantumlib/OpenFermion-Psi4/blob/master/openfermionpsi4/_psi4_conversion_functions.py#L97
+def parse_amplitudes_from_psi4_ccsd(
```
Comment (mstechly) [inline src/python/qepsi4/_ccsd_amplitude_parser.py:22]: Can we split this function into multiple smaller ones?
It's hard to follow and probably also hard to test the way it's now.

**Review by mstechly — CHANGES_REQUESTED** [review]

File: src/python/qepsi4/_ccsd_amplitude_parser.py (hunk 2 above)
Comment (Shiro-Raven) [reply src/python/qepsi4/_ccsd_amplitude_parser.py:23]: @mstechly Done, can you please check how it looks now?

File: src/python/qepsi4/_ccsd_amplitude_parser.py [hunk 3]
```diff
@@ -301,3 +232,77 @@ def beta_unoccupied(i):
     fermion_generator = s_generator + d_generator
 
     return fermion_generator
+
+
+def extract_T_indices(output_buffer, get_mp2_amplitudes, mp2_line, ccsd_line):
```
Comment (mstechly) [inline src/python/qepsi4/_ccsd_amplitude_parser.py:237]: I think you can make these helper methods private.

Comment (mstechly) [reply src/python/qepsi4/_ccsd_amplitude_parser.py:237]: Also, it lacks typing, would be nice to have that if possible

File: src/python/qepsi4/_ccsd_amplitude_parser.py (hunk 2 above)
Comment (mstechly) [reply src/python/qepsi4/_ccsd_amplitude_parser.py:23]: It looks better, but is still hard to follow. If you think you can extract something more that would be great, if not that's fine.

Comment (mstechly) [reply src/python/qepsi4/_ccsd_amplitude_parser.py:23]: But please add typing :)

File: tests/qepsi4/_ccsd_amplitude_parser_test.py [hunk 4]
```diff
@@ -0,0 +1,290 @@
+import pytest
+from qepsi4 import run_psi4, parse_amplitudes_from_psi4_ccsd
+from openfermion import FermionOperator
+
+
+def get_expected_mp2_fop():
+    result_fop = FermionOperator()
+    result_fop += -0.0193104873 * FermionOperator("4^ 0 5^ 1")
```
Comment (mstechly) [inline tests/qepsi4/_ccsd_amplitude_parser_test.py:8]: I assume that we cannot create a test for an example that is simple enough that we could calculate it on paper? Cause it looks to me that these are just generated by the code you're trying to test, so they don't really check whether the code works properly, it just checks whether the code does not change.

**Review by mstechly — APPROVED** [review]
